### ISSUE TITLE
Guitar tap note placement seems inaccurate

### ISSUE THREAD

**Issue by NewCreature** [NewCreature/EOF#27]
_From [raynebc](https://code.google.com/u/raynebc/) on May 10, 2010 13:37:11_

Guitar tap places notes that are noticeably later than they should be,
regardless of grid snap. For example, using guitar tap to place notes on a
beat synced chart, all notes are found to be placed out of sync during
playback.

I can't remember if this same issue occurs in Guitar Strum input method.

_Original issue: http://code.google.com/p/editor-on-fire/issues/detail?id=27_

**Comment by NewCreature** [comment]
_From [raynebc](https://code.google.com/u/raynebc/) on May 12, 2010 10:36:45_

This issue reminded me of a recent fix I made with the lyric preview timing.  It may
be worth checking if the AV Delay is being accounted for in the timestamp at which
notes placed with guitar tap/strum are placed.  If it isn't, it would explain why
they are all placed out of sync with the chart.

**Comment by NewCreature** [comment]
_From [raynebc](https://code.google.com/u/raynebc/) on May 12, 2010 10:51:39_

The above example patched in r36

**Comment by NewCreature** [comment]
_From [xander4j...@yahoo.com](https://code.google.com/u/111302640723734240985/) on May 12, 2010 11:08:44_

AV Delay is accounted for.

**Comment by NewCreature** [comment]
_From [xander4j...@yahoo.com](https://code.google.com/u/111302640723734240985/) on May 12, 2010 17:18:14_

I can't reproduce this issue. The notes are always entered exactly where the position 
line is. At first they appeared early (but at the position line) but I adjust the AV 
delay and now they are placed correctly.

**Comment by NewCreature** [comment]
_From [raynebc](https://code.google.com/u/raynebc/) on May 12, 2010 17:37:07_

I'll have to do some more testing to try to identify this, but I cannot get guitar
tap/strum to place notes anywhere near the beatlines (within .05 seconds) on a beat
synced chart unless I intentionally hit the keyboard early.

**Comment by NewCreature** [comment]
_From [xander4j...@yahoo.com](https://code.google.com/u/111302640723734240985/) on May 13, 2010 12:14:55_

Just a thought, but maybe the controller has a delay. If so, perhaps I would need to 
add another setting for controller delay.

**Comment by NewCreature** [comment]
_From [raynebc](https://code.google.com/u/raynebc/) on May 13, 2010 12:25:31_

I've reproduced this off-timing problem with the regular keyboard, but I believe
regular USB game controllers had the same issue.  Historically, FoF users have
related a horrible controller delay to MIDI based controllers like an electronic drumset.

**Comment by NewCreature** [comment]
_From [xander4j...@yahoo.com](https://code.google.com/u/111302640723734240985/) on May 13, 2010 12:45:45_

Code-wise I can't see any way the notes would be placed incorrectly. The only 
possibility I see is if there were some delay between when you pressed a button/key 
and when EOF registered that button/key as pressed.

If you watch the screen and can visibly see when the note appears, you should see it 
appear exactly at the position line every time. The only time this would not be the 
case is if the program was skipping frames. Under that circumstance you won't be 
able to use guitar strum/tap modes anyway because it would be unreliable.

Just because something is hooked up via a wire doesn't mean it is instantly 
responsive. There's also the issue with wireless controllers and even drivers 
(joy2key comes to mind).

The bottom line is either the AV delay is not exactly correct or there is some delay 
in the controller. I think adding an option for controller delay (one for drums and 
one for guitar) would be the only way to solve this.

**Comment by NewCreature** [comment]
_From [raynebc](https://code.google.com/u/raynebc/) on May 13, 2010 13:10:24_

I'm pretty certain my AV delay is much more accurate than how far off the notes are
being placed, but without a visual cue for AV delay, it can't be set as precisely as
controller input would need.

Going about it from your other proposed way (controller delay), if you could add a
Rock Band style calibration to press a button (ie. strum) in time with a steady
metronome (and perform the mean between the time differences), it could be a good way
to find the best delay to compensate with.  This controller delay could potentially
be altered in EOF's menu in the way that AV delay currently is able to be adjusted
manually.

**Comment by NewCreature** [comment]
_From [xander4j...@yahoo.com](https://code.google.com/u/111302640723734240985/) on May 14, 2010 08:15:43_

**Status:** Started

**Comment by NewCreature** [comment]
_From [raynebc](https://code.google.com/u/raynebc/) on May 14, 2010 08:48:21_

It's possible that the platform-specific playback code will effect this.  I'll be
able to test the changes from r111 on this issue when I get home.

**Comment by NewCreature** [comment]
_From [xander4j...@yahoo.com](https://code.google.com/u/111302640723734240985/) on May 14, 2010 09:40:04_

**Status:** Fixed

**Comment by NewCreature** [comment]
_From [raynebc](https://code.google.com/u/raynebc/) on May 15, 2010 01:13:28_

The controller delay seems to work well, after setting a drum delay of 120ms, tapping
the keyboard in time with the song places notes on the beat lines.  I'll work on the
documentation for this feature.
